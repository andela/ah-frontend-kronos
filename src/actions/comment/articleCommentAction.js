import axios from 'axios';
import { toastSuccess, toastFailure } from '../../utils/toast';
import 'react-toastify/dist/ReactToastify.css';
import { articleCommentAction } from '../actionTypes';
import URLS from '../../urlConnections';

export const fetchComment = slug => dispatch => axios.get(`${URLS.BACKEND_URL}/api/articles/${slug}/comments/`, {
  headers: {
    'content-type': 'application/json',
    Authorization: `Bearer ${sessionStorage.getItem('token')}`,
  },
})
  .then((response) => {
    dispatch({
      type: articleCommentAction.REQUEST_COMMENT_SUCCESS,
      payload: response.data.comment.comments,
    });
  }).catch((error) => {
    dispatch({
      type: articleCommentAction.REQUEST_COMMENT_FAILURE,
      payload: error.response.data.comment.error,
    });
  });

export const postComment = (slug, comment) => dispatch => axios.post(`${URLS.BACKEND_URL}/api/articles/${slug}/comments/`, comment, {
  headers: {
    'content-type': 'application/json',
    Authorization: `Bearer ${sessionStorage.getItem('token')}`,
  },
})
  .then((response) => {
    dispatch({
      type: articleCommentAction.CREATE_COMMENT_SUCCESS,
      payload: response.data.comment,
    });
    toastSuccess('Comment successfully created', 'A');
  }).catch((error) => {
    dispatch({
      type: articleCommentAction.CREATE_COMMENT_FAILURE,
      payload: error.response.data.comment.error,
    });
    toastFailure('Failed to create comment, try again', 'A');
  });

export const deleteComment = (slug, commentId) => dispatch => axios.delete(`${URLS.BACKEND_URL}/api/articles/${slug}/comments/${commentId}`, {
  headers: {
    'content-type': 'application/json',
    Authorization: `Bearer ${sessionStorage.getItem('token')}`,
  },
})
  .then((response) => {
    dispatch({
      type: articleCommentAction.DELETE_COMMENT_SUCCESS,
      payload: response.data.comment.message,
    });
    toastSuccess('Comment successfully deleted', 'A');
  }).catch((error) => {
    dispatch({
      type: articleCommentAction.DELETE_COMMENT_FAILURE,
      payload: error.response.data.comment.error,
    });
    toastFailure('Failed to delete comment, its not you comment!', 'A');
  });


export const editComment = (slug, commentId, commentUpdate) => dispatch => axios.patch(`${URLS.BACKEND_URL}/api/articles/${slug}/comments/${commentId}`, commentUpdate, {
  headers: {
    'content-type': 'application/json',
    Authorization: `Bearer ${sessionStorage.getItem('token')}`,
  },
})
  .then((response) => {
    dispatch({
      type: articleCommentAction.EDIT_COMMENT_SUCCESS,
      payload: response.data.comment,
    });
    toastSuccess('Comment successfully updated!', 'A');
  }).catch((error) => {
    dispatch({
      type: articleCommentAction.EDIT_COMMENT_FAILURE,
      payload: error.response.data.comment,
    });
    toastFailure('Failed to edit comment, its not you comment!', 'A');
  });
