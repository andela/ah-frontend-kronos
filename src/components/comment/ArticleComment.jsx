/* eslint-disable react/void-dom-elements-no-children */
/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable jsx-a11y/no-static-element-interactions */
/* eslint-disable jsx-a11y/anchor-is-valid */

import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import '../../assets/scss/Comment.scss';

export class ArticleComment extends Component {
  constructor(props) {
    super(props);
    this.state = {
      commentId: '',
      commentBody: '',
      commentUpdate: '',
      btnChange: false,
      outPutBtn: false,
    };
  }

  componentDidMount() {
    const { slug, getComment } = this.props;
    getComment(slug);
  }

  componentWillReceiveProps(nextProps) {
    const { slug, getComment } = nextProps;
    getComment(slug);
  }

  handleChange = (event) => {
    this.setState({
      [event.target.name]: event.target.value,
    });
  }

  handleDelete = (commentId) => {
    const { slug, removeComment } = this.props;
    removeComment(slug, commentId);
  }

  handleUpdateButton = (commentId) => {
    const { comments: { comments } } = this.props;
    const [commentChange] = comments.filter(comment => comment.id === commentId);
    this.setState({ btnChange: true, commentId, commentUpdate: commentChange.comment_body });
  }

  handleCancelEdit = () => {
    this.setState({ btnChange: false });
  }

  handleValidation = (event) => {
    event.preventDefault();
    this.setState({ outPutBtn: true });
  }

  handleUpdate = (commentId) => {
    const { slug, updateComment } = this.props;
    const { commentUpdate } = this.state;
    const comment = { comment_body: commentUpdate };
    updateComment(slug, commentId, comment);
    this.setState({ btnChange: false });
  }

  handleSubmit = (event) => {
    event.preventDefault();
    const { slug, createComment } = this.props;
    const { commentBody } = this.state;
    const comment = { comment_body: commentBody };
    createComment(slug, comment);
    this.setState({ commentBody: '' });
  }

  render() {
    const { comments: { comments } } = this.props;
    const {
      btnChange, outPutBtn, commentBody, commentUpdate, commentId,
    } = this.state;

    return (
      <React.Fragment>
        <div className="form-comment container">
          <div className="container">
            <div className="container">
              <h5>Comments </h5>
              {comments.map(comment => (
                <div className="card" key={comment.id}>
                  <div className="card-body">
                    <div className="row card-row">
                      <div className="col-md-2">{!comment.author.image || comment.author.image === 'url' ? (<img src="https://image.ibb.co/jw55Ex/def_face.jpg" className="img irounded-circle img-fluid fluid" alt="profile" />) : (<img src={comment.author.image} className="rounded-circle img img-fluid fluid" alt="profile" />)}</div>
                      <div className="col-md-10">
                        <p><a className="float-left"><strong className="comment-author">{comment.author.username}</strong></a></p>
                        <div className="clearfix" />
                        {btnChange === true && commentId === comment.id ? (<textarea className="form-control" value={commentUpdate} onChange={this.handleChange} onBlur={this.handleValidation} name="commentUpdate" rows="5" id="comment" />)
                          : (<p type="text" className="text-body">{comment.comment_body}</p>)}
                        <p className="time-text text-secondary">{`Created_at: ${(new Date(comment.created_at).toLocaleString())}`}</p>
                        {sessionStorage.getItem('isLoggedIn') === 'true' ? (
                          <p>
                            {btnChange === true && commentId === comment.id ? (
                              <a className="float-right btn">
                                <i className="fa fa-paper-plane font-btn" onClick={() => this.handleUpdate(comment.id)} />
                                <i className="fa fa-window-close font-btn" onClick={this.handleCancelEdit} />
                              </a>
                            ) : (
                              <a className="float-right btn">
                                <i className="fa fa-edit font-btn" onClick={() => this.handleUpdateButton(comment.id)} />
                                <i className="fa fa-trash font-btn" onClick={() => this.handleDelete(comment.id)} />
                              </a>
                            )}
                          </p>
                        ) : null}
                      </div>
                    </div>
                  </div>
                </div>
              ))}
            </div>
            {sessionStorage.getItem('isLoggedIn') === 'true' ? (
              <div className="container">
                <div className="form-group">
                  <textarea className="form-control" placeholder="Write your comment  here" name="commentBody" value={commentBody} onChange={this.handleChange} onKeyUp={this.handleValidation} rows="5" id="comment" />
                </div>
                {outPutBtn === true ? (
                  <div>
                    <button type="submit" className="btn-primary" onClick={this.handleSubmit}>Submit</button>
                  </div>
                ) : null }
              </div>
            ) : null}
          </div>
        </div>
      </React.Fragment>
    );
  }
}

export const mapStateToProps = state => ({
  comments: state,
});

ArticleComment.propTypes = {
  comments: PropTypes.shape({}).isRequired,
  getComment: PropTypes.func,
  createComment: PropTypes.func,
  updateComment: PropTypes.func,
  removeComment: PropTypes.func,
  slug: PropTypes.string.isRequired,
};

ArticleComment.defaultProps = {
  getComment: () => {},
  createComment: () => {},
  updateComment: () => {},
  removeComment: () => {},
};

export default connect(
  mapStateToProps,
)(ArticleComment);
