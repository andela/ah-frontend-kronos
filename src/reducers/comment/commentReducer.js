import { articleCommentAction } from '../../actions/actionTypes';
import { initialCommentState } from '../initialState';


export default function (state = initialCommentState, action) {
  switch (action.type) {
    case articleCommentAction.REQUEST_COMMENT_SUCCESS:
      return {
        ...state,
        comments: action.payload,
      };
    case articleCommentAction.REQUEST_COMMENT_FAILURE:
      return {
        ...state,
        error: action.payload,
      };
    case articleCommentAction.DELETE_COMMENT_SUCCESS:
      return {
        ...state,
        delComment: action.payload,
      };
    case articleCommentAction.DELETE_COMMENT_FAILURE:
      return {
        ...state,
        error: action.payload,
      };
    case articleCommentAction.EDIT_COMMENT_SUCCESS:
      return {
        ...state,
        editComments: action.payload,
      };
    case articleCommentAction.EDIT_COMMENT_FAILURE:
      return {
        ...state,
        error: action.payload,
      };
    case articleCommentAction.CREATE_COMMENT_SUCCESS:
      return {
        ...state,
        createComments: action.payload,
      };
    case articleCommentAction.CREATE_COMMENT_FAILURE:
      return {
        ...state,
        error: action.payload,
      };
    case articleCommentAction.CREATE_COMMENT_ERROR:
      return {
        ...state,
        error: action.payload,
      };
    default:
      return state;
  }
}
