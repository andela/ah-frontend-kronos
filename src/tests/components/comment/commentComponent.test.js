import React from 'react';
import { shallow } from 'enzyme';
import { ArticleComment, mapStateToProps } from '../../../components/comment/ArticleComment';

// test that ArticleComment component
describe('<ArticleComment />', () => {
  let wrapper;
  const props = {
    comments: {
      comments: [{
        id: 1,
        comment_body: 'Programing is my style',
        created_at: '01/12/2091',
        updated_at: '2019-07-24T16:12:46.624323Z',
        author: {
          image: 'url',
          username: 'testmanuel',
          first_name: 'Dominic',
          last_name: 'Manuel',
          created_at: '2019-07-17T08:36:32.249335Z',
          bio: 'Software developer',
          following: true,
          date_of_birth: '2019-07-18',
        },
        article: 83,
        replies: [],
        comment_on_text: null,
        comment_on_start: null,
        comment_on_end: null,
        user_like_status: null,
        likes_count: null,
      }],
    },
    slug: 'test',
    createComment: jest.fn(),
    getComment: jest.fn(),
    updateComment: jest.fn(),
    removeComment: jest.fn(),
    history: {
      push: jest.fn(),
    },
  };

  beforeAll(() => {
    wrapper = shallow(<ArticleComment {...props} />);
  });

  it('renders without fail', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('it should map state to props', () => {
    const initialState = {
      comments: { comments: {} },
    };
    const state = mapStateToProps(initialState);
    expect(state.comments).toEqual({ comments: { comments: {} } });
  });
  it('should handle change', () => {
    const event = {
      preventDefault: jest.fn(),
      target: {
        name: 'commentBody',
        value: 'Programing is my style',
      },
    };
    const wrapperInst = wrapper.instance();
    wrapperInst.handleChange(event);
    wrapperInst.handleDelete(event);
    wrapperInst.handleCancelEdit();
    wrapperInst.handleValidation(event);
    wrapperInst.handleUpdate(event);
    wrapperInst.handleSubmit(event);
    const expectedData = {
      btnChange: false,
      commentBody: 'Programing is my style',
      commentId: {
        preventDefault: jest.fn(),
        target: {
          name: 'commentBody',
          value: 'Programing is my style',
        },
      },
      commentUpdate: '',
      outPutBtn: true,
      slug: undefined,
    };
    expect(JSON.stringify(
      wrapperInst.state.commentUpdate,
    )).toBe(JSON.stringify(expectedData.commentUpdate));
  });
});
