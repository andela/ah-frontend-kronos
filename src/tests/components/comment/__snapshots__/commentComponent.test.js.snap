// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ArticleComment /> renders without fail 1`] = `ShallowWrapper {}`;
